### ISSUE TITLE
Configure markdown files to open in preview mode by default

### ISSUE THREAD

**Issue by cbochs** [zed-industries/zed#10966]
### Check for existing issues

- [X] Completed

### Describe the feature

I know you can open a file at a specific line with the `zed` cli with:

```
zed README.md:111
```

It would be great if you could open a markdown file could be opened (by default) in "preview" mode instead of the editable document? In VSCode this can be accomplished with `workspace.editorAssociations`:

```json
  "workbench.editorAssociations": {
    "*.ipynb": "jupyter-notebook",
    "*.md": "vscode.markdown.preview.editor"
  },
```

**Comment by ghuh** [comment]
This feature is not yet implemented, but still really should be.

**Comment by bobbymannino** [comment]
i think this is a great idea, but i think zed could do better then vscode, what if they're was an option for 3 states for markdown to open as: `code` (as default), `preview` and `preview_to_side` where they open in split screen. i think this would be one of those features that separates zed even more from any other editor

**Comment by aevsai** [comment]
Hey everyone, I drafted a solution to see if there is quick way to add this feature.

I added a workspace setting `file_preview_modes` that allows users to automatically dispatch actions when opening files matching specific patterns.

### Configuration
```json
{
  "workspace": {
    "file_preview_modes": [
      {
        "filter": "*.md",
        "command": "markdown::OpenPreview"
      },
      {
        "filter": "*.svg", 
        "command": "svg::OpenPreviewToTheSide"
      }
    ]
  }
}
```

### Implementation Details

**Location:** `crates/workspace/src/workspace.rs` in `open_path_preview()` method

**Flow:**
1. User opens a file (via project panel, CLI, file finder, etc.)
2. File opens normally as editor in `open_path_preview()`
3. After item is created, we check settings for matching patterns
4. If pattern matches, we build and dispatch the configured action (e.g., `markdown::OpenPreview`)
5. The action handler executes and creates the preview view

**Code Structure:**
- `FilePreviewModeRule` struct: `{ filter: String, command: String }`
- Pattern matcher: `crates/workspace/src/file_preview_matcher.rs`
- Settings integration: `workspace.file_preview_modes: Vec<FilePreviewModeRule>`
- Interception point: After `pane.open_item()` completes in `open_path_preview()`

### Current Status

- Settings are correctly loaded
- Pattern matching works (exact filename, wildcards, path patterns)
- Actions are built and dispatched successfully
- Works perfectly when opening from **project panel**

**Issues:**

#### Action Dispatch from Markdown Links Fails
**Symptom:** When clicking a markdown link in preview, file opens as editor but preview doesn't open.

**Root Cause:** Action dispatch timing problem. The action is dispatched immediately after the editor item is created, but the action handler needs the editor to be "active" in the workspace context.

Action dispatches successfully, but `MarkdownPreviewView::register()` action handler checks:
```rust
if let Some(editor) = Self::resolve_active_item_as_markdown_editor(workspace, cx) {
    // Create preview...
}
```

The editor isn't "active" yet, so the handler returns early without creating the preview.

**Why it works from project panel:** Different timing - the item has time to become active before action is dispatched.

#### Redundant Editor Creation
**Current behavior:**
1. File opens as editor
2. Action creates preview
3. Result: Editor + Preview both open

**Expected behavior:**
- Only preview should open (no editor first)

**Impact:** 
- Extra pane/tab created
- On close, focus returns to editor instead of just closing
- User sees brief flash of editor before preview appears
- Not intuitive UX


So basically if we want it to work as actual "file opened in mode" and not "action dispatched after file opened" then we need to refactor action OpenPreview and other to work without active editor tab. Otherwise we can call action after tab creation with defer.

I would be glad If someone from maintainers could help me understand which approach is better aligned with Zed

P.s. naming would be changed to be similar to vscode

**Comment by aevsai** [comment]
Just sent pull request with what looks like nice and correct implementation to me.

https://github.com/user-attachments/assets/13c6d403-4c4a-4863-927b-4d6626b7281f

**Comment by atlasstaking** [comment]
Wow this would awesome indeed
Claude does the md wrting now and we do only the reading indeed :-)

**Comment by Cheirnomus** [comment]
I need this yesterday and everyday!

**Comment by Cheirnomus** [comment]
In the meantime, I already wrote a MD viewer in a HTML page or also just convert all MD files to PDFs via a script. Why is this simple feature request for a everyday use case taking so long?

**Comment by amyin** [comment]
Would also love this feature! Whenever I have my ai do an analysis, i ask it to pop the completed file open in ZED and I would love to not have to click the eyeball icon to see the markdown displayed nicely 🙏

**Comment by shivamashtikar** [comment]
I was thinking if there's any way to listen to which file is opened and toggle Markdown preview mode if it is a markdown file, if this feature is tedious enough to implement

**Comment by Evan-Luo-Engnieer** [comment]
+1 — waiting for this feature. Following PR #48733.

**Comment by SomeoneToIgnore** [comment]
https://github.com/zed-industries/zed/pull/48733 has a good overview of what needs to be done generally for the feature.
Unfortunately, the PR misses the CSV preview and does not work — fixing that and actually testing the app before submitting a PR should be the last two things left to do here.

**Comment by lafllamme** [comment]
Push, Im really annoyed by always double clicking my whole notes!

**Comment by g-kartik** [comment]
Selected text in preview mode in markdown file cannot be directly sent to agent thread via `Ctrl + Shift + >`. This is very useful in citing text directly to the agent, instead of going back and forth between edit and preview mode.

**Comment by SomeoneToIgnore** [comment]
https://github.com/zed-industries/zed/pull/57010 was a nice increment since the last attempt: in that PR there is a code that covers all kinds of items we can preview, and actually shows the preview!

Yet, it has issues with usability and bugs when preview is not re-opened: 

https://github.com/user-attachments/assets/ffaf1177-6a3b-4890-ad46-471d992bca08

For comparison, VSCode:

https://github.com/user-attachments/assets/f063a2ad-576d-4d71-aa50-abfdc98b3d7b

I see at least the following UX discrepancies:

* Auto preview actually opens when I open another file
* Auto preview does not get the focus, it's the editor that keeps it
* Auto preview gets replaced instead of duplicating

I suspect there could be more, so the next iteration could concentrate on having that UX polished: for that, one needs to run Zed and another editor and compare cases; write a few more tests on various scenarios (at least one); improve the preview behavior and potentially extract the common code.

**Comment by cug** [comment]
I'd love to have this feature as well. I use it in a way that I have the Agent work on or generate MD files, which I generally only want to view.

**Comment by atqamz** [comment]
I continue the prev work from rejected PR, please kindly review https://github.com/zed-industries/zed/pull/58752
any feedback matter to me

video demo will come later, thanks

**Comment by atqamz** [comment]
just uploaded the demo vid on PR's comment https://github.com/zed-industries/zed/pull/58752#issuecomment-4641615186

**Comment by derekneely** [comment]
Is this coming anytime soon? Pretty much can't switch to Zed until this feature is implemented. Would slow us down to much otherwise.

**Comment by lafllamme** [comment]
Pushhhhh! I'm literally crying, everytime I have to double click my whole azz docs folder

**Comment by shepitchakpavlo** [comment]
upvoting 🗳️🗳️🗳️

**Comment by m0rg5** [comment]
I just set a custom keystroke to open md preview, another to hide it, would like the option for md preview to open by default (and then would use keystroke to raw md if/when needed).

Why was this rejected?

**Comment by pletinsky** [comment]
Nobody writes md files directly in the brave new world. We use AI to do it now. Reading md file in edit mode by default does not make sense anymore. Pls implement the feature.

**Comment by lafllamme** [comment]
Any update on when I can finally stop smashing CMD + SHIFT + V? 😄

Also, is there anything we can do to help move this forward or get it merged? Happy to contribute if needed.

**Comment by kinoute** [comment]
That would be great to have this indeed.

**Comment by Greenheart** [comment]
Not tested it yet, but based on reading the release notes, maybe this issue is (fully/partially) resolved by https://github.com/zed-industries/zed/pull/63462?

**Comment by n-berilov** [comment]
Yes, I think this new option should be used to decide if we open in preview or raw.

**Comment by ghuh** [comment]
Personally, I'd like it to open in split preview so I can both edit the Markdown and see the preview window (i.e. `cmd+k v`) like IntelliJ does.  So https://github.com/zed-industries/zed/pull/63462 doesn't solve the issue.

**Comment by maphew** [comment]
This is implemented in 1.20.1, and it's great! but has the flaw of there being no visible way to switch to edit mode.

update: `ctrl-shift-V` toggles the modes
